DAOS-13327 test: Get the engine_per_host from YAML file.

rpadma2 opened the pull request:
Summary: Use the engine_per_host information from the YAML file rather than setting the engines to 2 in the test.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

rpadma2 committed:
DAOS-13327 test: Get the engine_per_host from YAML file.
Skip-unit-tests: true
Skip-fault-injection-test: true

Summary: Use the engine_per_host information from the
YAML file rather than setting the engines to 2 in the test.

Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-12113/1/testReport/(root)/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/2/execution/node/778/log

phender reviewed (changes requested):
We should make use of the `DaosServerManager.engines` and `DaosServerManager.ranks` properties to get the total number of engines and rank numbers.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/4/execution/node/430/log

rpadma2 committed:
DAOS-13327 test: Merge with master fix conflict.

Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

rpadma2 commented on src/tests/ftest/nvme/pool_exclude.py:
```diff
@@ -61,7 +64,7 @@ def run_nvme_pool_exclude(self, num_pool, oclass=None):
 
         # Exclude rank :  ranks other than rank 0.
-        exclude_servers = len(self.hostlist_servers) * 2
+        exclude_servers = len(self.hostlist_servers) * self.engine_count
         rank_list = list(range(1, exclude_servers))
```
Done

rpadma2 commented on src/tests/ftest/nvme/pool_extend.py:
```diff
@@ -46,7 +49,7 @@ def run_nvme_pool_extend(self, num_pool, oclass=None):
 
         daos_command = self.get_daos_command()
-        total_servers = len(self.hostlist_servers) * 2
+        total_servers = len(self.hostlist_servers) * self.engine_count
```
Done.

rpadma2 commented on src/tests/ftest/osa/online_drain.py:
```diff
@@ -49,7 +52,7 @@ def run_online_drain_test(self, num_pool, oclass=None, app_name="ior"):
             oclass = self.ior_cmd.dfs_oclass.value
         test_seq = self.ior_test_sequence[0]
-        drain_servers = (len(self.hostlist_servers) * 2) - 1
+        drain_servers = (len(self.hostlist_servers) * self.engine_count) - 1
```
Done.

rpadma2 commented on src/tests/ftest/osa/online_parallel_test.py:
```diff
@@ -93,7 +96,7 @@ def run_online_parallel_test(self, num_pool, racer=False):
 
         # Exclude target : random two targets  (target idx : 0-7)
-        exc = random.randint(0, 6)  # nosec
+        exc = random.randint(0, (self.hostlist_servers * self.engine_count))  # nosec
         target_list.append(exc)
         target_list.append(exc + 1)
```
Done.

rpadma2 commented on src/tests/ftest/osa/online_parallel_test.py:
```diff
@@ -36,6 +36,9 @@ def setUp(self):
         self.ior_dfs_oclass = self.params.get("obj_class", '/run/ior/iorflags/*')
         # Recreate the client hostfile without slots defined
         self.hostfile_clients = write_host_file(self.hostlist_clients, self.workdir, None)
+        # Get the engine count
+        self.engine_count = self.params.get("engines_per_host", '/run/server_config/*',
+                                            default=1)
```
REmoved.

rpadma2 commented on src/tests/ftest/osa/online_reintegration.py:
```diff
@@ -32,6 +32,9 @@ def setUp(self):
         self.daos_command = DaosCommand(self.bin)
         self.ior_test_sequence = self.params.get("ior_test_sequence", '/run/ior/iorflags/*')
         self.test_oclass = self.params.get("oclass", '/run/test_obj_class/*')
+        # Get the engine count
+        self.engine_count = self.params.get("engines_per_host", '/run/server_config/*',
+                                            default=1)
```
Removed

rpadma2 commented on src/tests/ftest/osa/online_reintegration.py:
```diff
@@ -60,7 +63,7 @@ def run_online_reintegration_test(self, num_pool, racer=False, server_boot=False
         # Create a pool
         pool = {}
-        exclude_servers = (len(self.hostlist_servers) * 2) - 1
+        exclude_servers = (len(self.hostlist_servers) * self.engine_count) - 1
 
         # Exclude one rank : other than rank 0.
         rank = random.randint(1, exclude_servers)  # nosec
```
Done.

rpadma2 committed:
DAOS-13327 test: Review updates
Test-tag: NvmePoolExclude NvmePoolExtend OSAOnlineDrain OSAOnlineReintegration OSAOnlineParallelTest OSAOfflineParallelTest

Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/5/execution/node/145/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/5/execution/node/1251/log

rpadma2 committed:
DAOS-13327 test: Merge branch 'master' into rpadma2/daos_13327

Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/6/execution/node/139/log

rpadma2 committed:
DAOS-13327 test: Fix pool extend pylint issue.
Test-tag: NvmePoolExclude NvmePoolExtend OSAOfflineDrain OSAOfflineExtend OSAOfflineReintegration OSAOnlineDrain OSAOnlineExtend OSAOnlineReintegration

Required-githooks: true

Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/7/execution/node/140/log

rpadma2 commented:
Started this ticket long time before with @phender input... Forgot to follow-up on this PR for a quite sometime. Sigh :-(

daltonbohning commented on src/tests/ftest/nvme/pool_exclude.py:
```diff
@@ -59,8 +59,8 @@ def run_nvme_pool_exclude(self, num_pool, oclass=None):
         # Exclude rank :  ranks other than rank 0.
-        exclude_servers = len(self.hostlist_servers) * 2
-        rank_list = list(range(1, exclude_servers))
+        rank_list = list(self.server_managers[0].ranks.keys())
+        rank_list.remove(0)
```
FYI that special handling for rank 0 is starting to be less of a guarantee. E.g. it isn't guaranteed that rank 0 is the pool leader

daltonbohning commented on src/tests/ftest/nvme/pool_extend.py:
```diff
@@ -45,7 +45,7 @@ def run_nvme_pool_extend(self, num_pool, oclass=None):
         ior_test_sequence = self.params.get("ior_test_sequence", '/run/ior/iorflags/*')
 
-        total_servers = len(self.hostlist_servers) * 2
+        total_servers = self.server_managers[0].engines()
```
FYI this is a property so this works too
```suggestion
        total_servers = self.server_managers[0].engines
```

daltonbohning commented on src/tests/ftest/osa/offline_drain.py:
```diff
@@ -29,7 +29,10 @@ def setUp(self):
             "ior_test_sequence", '/run/ior/iorflags/*')
         # Recreate the client hostfile without slots defined
-        self.hostfile_clients = write_host_file(self.hostlist_clients, self.workdir)
+        self.hostfile_clients = write_host_file(self.hostlist_clients, self.workdir, None)
```
No need to pass `slots=None` anymore
```suggestion
        self.hostfile_clients = write_host_file(self.hostlist_clients, self.workdir)
```

This is probably a bad (minor) merge

daltonbohning commented on src/tests/ftest/osa/offline_drain.py:
```diff
@@ -48,7 +51,7 @@ def run_offline_drain_test(self, num_pool, data=False, oclass=None, pool_fillup=
 
         # Exclude target : random two targets  (target idx : 0-7)
-        exc = random.randint(0, 6)  # nosec
+        exc = random.randint(0, (self.hostlist_servers * self.engine_count))  # nosec
```
Pre-existing, but we should replace `random` with `self.random`. Then we don't need `# nosec`
```suggestion
        exc = self.random.randint(0, (self.hostlist_servers * self.engine_count))
```

daltonbohning commented on src/tests/ftest/osa/offline_parallel_test.py:
```diff
@@ -83,9 +86,8 @@ def run_offline_parallel_test(self, num_pool, data=False, oclass=None):
-        target_list.append(exc)
-        target_list.append(exc + 1)
+        targets = self.server_managers[0].get_config_value("targets")
+        target_list = random.sample(list(range(0, targets)), 2)  # nosec
```
```suggestion
        target_list = self.random.sample(list(range(0, targets)), 2)
```

daltonbohning commented on src/tests/ftest/osa/online_drain.py:
```diff
@@ -46,16 +46,14 @@ def run_online_drain_test(self, num_pool, oclass=None, app_name="ior"):
-        target_list.append(exc)
-        target_list.append(exc + 1)
+        targets = self.server_managers[0].get_config_value("targets")
+        target_list = random.sample(list(range(0, targets)), 2)  # nosec
```
```suggestion
        target_list = self.random.sample(list(range(0, targets)), 2)
```

daltonbohning commented on src/tests/ftest/osa/online_drain.py:
```diff
@@ -46,16 +46,14 @@ def run_online_drain_test(self, num_pool, oclass=None, app_name="ior"):
 
         # Drain one of the ranks (or server)
-        rank = random.randint(1, drain_servers)  # nosec
+        rank = random.choice(list(self.server_managers[0].ranks.keys())[1:])
```
```suggestion
        rank = self.random.choice(list(self.server_managers[0].ranks.keys())[1:])
```

daltonbohning commented on src/tests/ftest/osa/online_parallel_test.py:
```diff
@@ -93,9 +93,8 @@ def run_online_parallel_test(self, num_pool, racer=False):
-        target_list.append(exc)
-        target_list.append(exc + 1)
+        targets = self.server_managers[0].get_config_value("targets")
+        target_list = random.sample(list(range(0, targets)), 2)  # nosec
```
```suggestion
        target_list = self.random.sample(list(range(0, targets)), 2)
```

daltonbohning commented on src/tests/ftest/osa/online_reintegration.py:
```diff
@@ -58,10 +58,9 @@ def run_online_reintegration_test(self, num_pool, racer=False, server_boot=False
 
         # Exclude one rank : other than rank 0.
-        rank = random.randint(1, exclude_servers)  # nosec
+        rank = random.choice(list(self.server_managers[0].ranks.keys())[1:])  # nosec
```
```suggestion
        rank = self.random.choice(list(self.server_managers[0].ranks.keys())[1:])
```

daltonbohning reviewed (dismissed):
Just code maintenance if you need to push again

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/7/execution/node/1511/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/7/execution/node/1572/log

rpadma2 committed:
DAOS-13327 test: Merge branch 'master' into rpadma2/daos_13327

Signed-off-by: rpadma2 <ravindran.padmanabhan@hpe.com>

rpadma2 commented:
Sigh. I didn't realize this PR was approved long time before and I didn't follow up.... Merging with the master and running the testing.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/8/execution/node/1331/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12113/8/execution/node/1430/log
